openpype/pipeline/plugin_discover.py

```diff
@@ -0,0 +1,264 @@
+import os
+import inspect
+import traceback
+
+from openpype.lib.python_module_tools import (
+    modules_from_path,
+    classes_from_module,
+)
+
+
+class DiscoverResult:
+    """Hold result of publish plugins discovery.
+
+    Stores discovered plugins duplicated plugins and file paths which
+    crashed on execution of file.
+    """
+
+    def __init__(self):
+        self.plugins = []
+        self.crashed_file_paths = {}
+        self.duplicated_plugins = []
+        self.abstract_plugins = []
+        self.ignored_plugins = set()
+
+    def __iter__(self):
+        for plugin in self.plugins:
+            yield plugin
+
+    def __getitem__(self, item):
+        return self.plugins[item]
+
+    def __setitem__(self, item, value):
+        self.plugins[item] = value
+
+    def get_report(self, only_errors=True, exc_info=True, full_report=False):
+        lines = []
+        if not only_errors:
+            # Successfully discovered plugins
+            if self.plugins or full_report:
+                lines.append(
+                    "*** Discovered {} plugins".format(len(self.plugins))
+                )
+                for cls in self.plugins:
+                    lines.append("- {}".format(cls.__class__.__name__))
+
+            # Plugin that were defined to be ignored
+            if self.ignored_plugins or full_report:
+                lines.append("*** Ignored plugins {}".format(len(
+                    self.ignored_plugins
+                )))
+                for cls in self.ignored_plugins:
+                    lines.append("- {}".format(cls.__class__.__name__))
+
+        # Abstract classes
+        if self.abstract_plugins or full_report:
+            lines.append("*** Discovered {} abstract plugins".format(len(
+                self.abstract_plugins
+            )))
+            for cls in self.abstract_plugins:
+                lines.append("- {}".format(cls.__class__.__name__))
+
+        # Abstract classes
+        if self.duplicated_plugins or full_report:
+            lines.append("*** There were {} duplicated plugins".format(len(
+                self.duplicated_plugins
+            )))
+            for cls in self.duplicated_plugins:
+                lines.append("- {}".format(cls.__class__.__name__))
+
+        if self.crashed_file_paths or full_report:
+            lines.append("*** Failed to load {} files".format(len(
+                self.crashed_file_paths
+            )))
+            for path, exc_info_args in self.crashed_file_paths.items():
+                lines.append("- {}".format(path))
+                if exc_info:
+                    lines.append(10 * "*")
+                    lines.extend(traceback.format_exception(*exc_info_args))
+                    lines.append(10 * "*")
+
+        return "\n".join(lines)
+
+    def print_report(self, only_errors=True, exc_info=True):
+        report = self.get_report(only_errors, exc_info)
+        if report:
+            print(report)
```

Comment:
Shouldn't the printed report either be logged instead? Or put behind a verbosity flag?